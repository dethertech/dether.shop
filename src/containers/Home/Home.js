import React, { PureComponent } from 'react';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import { hasGoodNetwork as hasGoodNetworkHelper } from '../../reducers/app';

// Config

import config from '../../constants/config';

// Translate module

import tr from '../../translate';

// Redux

import {
  toggleTermsModal as toggleTermsModalAction,
  acceptTerms as acceptTermsAction,
} from '../../actions';
import {
  hasEnoughEthToAddShop,
  hasEnoughDthToAddShop,
} from '../../reducers/user';

// Components

import { Button, Message, Svg } from '../../components';
import { H1 } from '../../components/Headings';
import { Padding } from '../../components/Spaces';
import TermsValidation from './TermsValidation';

// Constants

import { hasSupportedBrowser } from '../../constants/browser';

// Styles

const Wrapper = styled.div`
  max-width: 40rem;
  margin: 0 auto;
  text-align: center;
`;

const Version = styled.span`
  font-size: 14px;
  font-weight: 400;
`;

const HomeImage = styled(Svg)`
  width: 80%;
`;

const Uppercase = styled.div`
  text-transform: uppercase;
`;

/**
 * Home containers
 * @extends PureComponent
 */
export class Home extends PureComponent {
  static propTypes = {
    isMetamaskInstalled: PropTypes.bool.isRequired,
    hasEnoughEth: PropTypes.bool.isRequired,
    hasEnoughDth: PropTypes.bool.isRequired,
    minEth: PropTypes.number.isRequired,
    minDth: PropTypes.string.isRequired,
    toggleTermsModal: PropTypes.func.isRequired,
    acceptTerms: PropTypes.func.isRequired,
    hasGoodNetwork: PropTypes.bool.isRequired,
  };

  constructor(props) {
    super(props);
    this.state = {
      checked: false,
      shake: 0,
    };
  }

  onClick = e => {
    const { acceptTerms } = this.props;
    if (this.checkTerms(e)) acceptTerms();
  };

  onCheck = ({ target: { checked } }) => {
    this.setState({ checked });
  };

  checkErrors = () => {
    const {
      isMetamaskInstalled,
      hasEnoughEth,
      hasEnoughDth,
      hasGoodNetwork,
      minEth,
      minDth,
    } = this.props;

    if (!hasSupportedBrowser) return tr('add.home.browser_not_supported');
    if (!isMetamaskInstalled)
      return tr('add.home.metamask_not_installed', {
        linkToMetamask: 'https://metamask.io/',
      });
    if (!hasGoodNetwork) return tr('add.home.wrong_network');
    if (!hasEnoughEth && !hasEnoughDth)
      return tr('add.home.not_enough_money', { minDth, minEth });

    if (!hasEnoughEth) return tr('add.home.not_enough_eth', { minEth });
    if (!hasEnoughDth) return tr('add.home.not_enough_dth', { minDth });
  };

  checkTerms = e => {
    const { checked, shake } = this.state;
    if (!checked) {
      e.preventDefault();
      this.setState({ shake: shake + 1 });
      return false;
    }
    return true;
  };

  render = () => {
    const { checked, shake } = this.state;
    const { toggleTermsModal } = this.props;
    const error = this.checkErrors();

    return (
      <React.Fragment>
        <Wrapper>
          <HomeImage type="SvgHome" />

          <H1>
            <Uppercase>{tr('add.home.title')}&nbsp;</Uppercase>
            <Version>v{config.appVersion}</Version>
          </H1>

          <Padding vertical="l">
            <span>{tr('add.home.desc')}</span>
          </Padding>

          {error ? (
            <Message theme="error" alignCenter>
              {error}
            </Message>
          ) : (
            <Padding all="l">
              <TermsValidation
                shake={shake}
                checked={checked}
                handleCheck={this.onCheck}
                toggleTermsModal={toggleTermsModal}
              />
              <Button fullWidth theme="primary" onClick={this.onClick}>
                {tr('add.home.bt_add')}
              </Button>
            </Padding>
          )}
        </Wrapper>
      </React.Fragment>
    );
  };
}

const mapStateToProps = ({ user, app }) => ({
  hasEnoughEth: hasEnoughEthToAddShop(user.balance, app.licencePrice),
  hasEnoughDth: hasEnoughDthToAddShop(user.balance, app.licencePrice),
  hasGoodNetwork: hasGoodNetworkHelper(app),
  isMetamaskInstalled: app.isMetamaskInstalled,
  isUserVerified: user.isCertified,
  minEth: config.gasPrice.simpleTransac,
  minDth: app.licencePrice,
});

const mapDispatchToProps = dispatch => ({
  toggleTermsModal: bindActionCreators(toggleTermsModalAction, dispatch),
  acceptTerms: bindActionCreators(acceptTermsAction, dispatch),
});

export default connect(mapStateToProps, mapDispatchToProps)(Home);
