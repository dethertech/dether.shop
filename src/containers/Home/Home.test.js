import React from 'react';
import { mount, shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import { Home } from './Home';

describe.skip('Containers::Home', () => {
  let wrapper;

  let props = {
    isMetamaskInstalled: true,
    hasEnoughEth: true,
    hasEnoughDth: true,
    minEth: 40,
    minDth: '100',
    toggleTermsModal: () => null,
    acceptTerms: () => null,
    hasGoodNetwork: true,
  }
  const component = props => <Home {...props} />

  it('should match shallow snapshop', () => {
    expect(shallow(component(props))).toMatchSnapshot('Home::shallow');
  });
  it('should match mounted snapshop', () => {
    expect(toJson(component(props))).toMatchSnapshot('Home::mount');
  });
});
