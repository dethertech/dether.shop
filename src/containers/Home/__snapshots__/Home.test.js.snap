// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home::mount 1`] = `undefined`;

exports[`Home::shallow 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Home
    acceptTerms={[Function]}
    hasEnoughDth={true}
    hasEnoughEth={true}
    hasGoodNetwork={true}
    isMetamaskInstalled={true}
    minDth="100"
    minEth={40}
    toggleTermsModal={[Function]}
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": Array [
        <styled.div>
          <Svg
                    type="SvgHome"
          />
          <styled.div
                    all={null}
                    bottom="m"
                    horizontal={null}
                    left={null}
                    right={null}
                    top={null}
                    vertical={null}
          >
                    <styled.h1>
                              CRYPTOCURRENCY FOR SHOPS
                               
                              <styled.span>
                                        v
                                        1.1.1
                              </styled.span>
                    </styled.h1>
          </styled.div>
          <styled.div
                    all={null}
                    bottom={null}
                    horizontal={null}
                    left={null}
                    right={null}
                    top={null}
                    vertical="l"
          >
                    <span>
                              Add your retail business on the Dether map. Expand your visibility in the Dether ecosystem of crypto buyers and sellers.
                    </span>
          </styled.div>
          <Message
                    alignCenter={true}
                    alignLeft={false}
                    theme="error"
                    withIcon={true}
          >
                    Your browser is not supported, Metamask is available on Google Chrome and Firefox only.
          </Message>
</styled.div>,
        <styled.div
          all={null}
          bottom={null}
          horizontal={null}
          left={null}
          right={null}
          top={null}
          vertical="s"
>
          <Footer
                    toggleTermsModal={[Function]}
          />
</styled.div>,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "children": Array [
            <Svg
              type="SvgHome"
/>,
            <styled.div
              all={null}
              bottom="m"
              horizontal={null}
              left={null}
              right={null}
              top={null}
              vertical={null}
>
              <styled.h1>
                            CRYPTOCURRENCY FOR SHOPS
                             
                            <styled.span>
                                          v
                                          1.1.1
                            </styled.span>
              </styled.h1>
</styled.div>,
            <styled.div
              all={null}
              bottom={null}
              horizontal={null}
              left={null}
              right={null}
              top={null}
              vertical="l"
>
              <span>
                            Add your retail business on the Dether map. Expand your visibility in the Dether ecosystem of crypto buyers and sellers.
              </span>
</styled.div>,
            <Message
              alignCenter={true}
              alignLeft={false}
              theme="error"
              withIcon={true}
>
              Your browser is not supported, Metamask is available on Google Chrome and Firefox only.
</Message>,
          ],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "type": "SvgHome",
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "all": null,
              "bottom": "m",
              "children": <styled.h1>
                CRYPTOCURRENCY FOR SHOPS
                 
                <styled.span>
                                v
                                1.1.1
                </styled.span>
</styled.h1>,
              "horizontal": null,
              "left": null,
              "right": null,
              "top": null,
              "vertical": null,
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "children": Array [
                  "CRYPTOCURRENCY FOR SHOPS",
                  " ",
                  <styled.span>
                    v
                    1.1.1
</styled.span>,
                ],
              },
              "ref": null,
              "rendered": Array [
                "CRYPTOCURRENCY FOR SHOPS",
                " ",
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "class",
                  "props": Object {
                    "children": Array [
                      "v",
                      "1.1.1",
                    ],
                  },
                  "ref": null,
                  "rendered": Array [
                    "v",
                    "1.1.1",
                  ],
                  "type": [Function],
                },
              ],
              "type": [Function],
            },
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "all": null,
              "bottom": null,
              "children": <span>
                Add your retail business on the Dether map. Expand your visibility in the Dether ecosystem of crypto buyers and sellers.
</span>,
              "horizontal": null,
              "left": null,
              "right": null,
              "top": null,
              "vertical": "l",
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Add your retail business on the Dether map. Expand your visibility in the Dether ecosystem of crypto buyers and sellers.",
              },
              "ref": null,
              "rendered": "Add your retail business on the Dether map. Expand your visibility in the Dether ecosystem of crypto buyers and sellers.",
              "type": "span",
            },
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "alignCenter": true,
              "alignLeft": false,
              "children": "Your browser is not supported, Metamask is available on Google Chrome and Firefox only.",
              "theme": "error",
              "withIcon": true,
            },
            "ref": null,
            "rendered": "Your browser is not supported, Metamask is available on Google Chrome and Firefox only.",
            "type": [Function],
          },
        ],
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "all": null,
          "bottom": null,
          "children": <Footer
            toggleTermsModal={[Function]}
/>,
          "horizontal": null,
          "left": null,
          "right": null,
          "top": null,
          "vertical": "s",
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "toggleTermsModal": [Function],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        "type": [Function],
      },
    ],
    "type": Symbol(react.fragment),
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": Array [
          <styled.div>
            <Svg
                        type="SvgHome"
            />
            <styled.div
                        all={null}
                        bottom="m"
                        horizontal={null}
                        left={null}
                        right={null}
                        top={null}
                        vertical={null}
            >
                        <styled.h1>
                                    CRYPTOCURRENCY FOR SHOPS
                                     
                                    <styled.span>
                                                v
                                                1.1.1
                                    </styled.span>
                        </styled.h1>
            </styled.div>
            <styled.div
                        all={null}
                        bottom={null}
                        horizontal={null}
                        left={null}
                        right={null}
                        top={null}
                        vertical="l"
            >
                        <span>
                                    Add your retail business on the Dether map. Expand your visibility in the Dether ecosystem of crypto buyers and sellers.
                        </span>
            </styled.div>
            <Message
                        alignCenter={true}
                        alignLeft={false}
                        theme="error"
                        withIcon={true}
            >
                        Your browser is not supported, Metamask is available on Google Chrome and Firefox only.
            </Message>
</styled.div>,
          <styled.div
            all={null}
            bottom={null}
            horizontal={null}
            left={null}
            right={null}
            top={null}
            vertical="s"
>
            <Footer
                        toggleTermsModal={[Function]}
            />
</styled.div>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": Array [
              <Svg
                type="SvgHome"
/>,
              <styled.div
                all={null}
                bottom="m"
                horizontal={null}
                left={null}
                right={null}
                top={null}
                vertical={null}
>
                <styled.h1>
                                CRYPTOCURRENCY FOR SHOPS
                                 
                                <styled.span>
                                                v
                                                1.1.1
                                </styled.span>
                </styled.h1>
</styled.div>,
              <styled.div
                all={null}
                bottom={null}
                horizontal={null}
                left={null}
                right={null}
                top={null}
                vertical="l"
>
                <span>
                                Add your retail business on the Dether map. Expand your visibility in the Dether ecosystem of crypto buyers and sellers.
                </span>
</styled.div>,
              <Message
                alignCenter={true}
                alignLeft={false}
                theme="error"
                withIcon={true}
>
                Your browser is not supported, Metamask is available on Google Chrome and Firefox only.
</Message>,
            ],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "type": "SvgHome",
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "all": null,
                "bottom": "m",
                "children": <styled.h1>
                  CRYPTOCURRENCY FOR SHOPS
                   
                  <styled.span>
                                    v
                                    1.1.1
                  </styled.span>
</styled.h1>,
                "horizontal": null,
                "left": null,
                "right": null,
                "top": null,
                "vertical": null,
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "class",
                "props": Object {
                  "children": Array [
                    "CRYPTOCURRENCY FOR SHOPS",
                    " ",
                    <styled.span>
                      v
                      1.1.1
</styled.span>,
                  ],
                },
                "ref": null,
                "rendered": Array [
                  "CRYPTOCURRENCY FOR SHOPS",
                  " ",
                  Object {
                    "instance": null,
                    "key": undefined,
                    "nodeType": "class",
                    "props": Object {
                      "children": Array [
                        "v",
                        "1.1.1",
                      ],
                    },
                    "ref": null,
                    "rendered": Array [
                      "v",
                      "1.1.1",
                    ],
                    "type": [Function],
                  },
                ],
                "type": [Function],
              },
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "all": null,
                "bottom": null,
                "children": <span>
                  Add your retail business on the Dether map. Expand your visibility in the Dether ecosystem of crypto buyers and sellers.
</span>,
                "horizontal": null,
                "left": null,
                "right": null,
                "top": null,
                "vertical": "l",
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": "Add your retail business on the Dether map. Expand your visibility in the Dether ecosystem of crypto buyers and sellers.",
                },
                "ref": null,
                "rendered": "Add your retail business on the Dether map. Expand your visibility in the Dether ecosystem of crypto buyers and sellers.",
                "type": "span",
              },
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "alignCenter": true,
                "alignLeft": false,
                "children": "Your browser is not supported, Metamask is available on Google Chrome and Firefox only.",
                "theme": "error",
                "withIcon": true,
              },
              "ref": null,
              "rendered": "Your browser is not supported, Metamask is available on Google Chrome and Firefox only.",
              "type": [Function],
            },
          ],
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "all": null,
            "bottom": null,
            "children": <Footer
              toggleTermsModal={[Function]}
/>,
            "horizontal": null,
            "left": null,
            "right": null,
            "top": null,
            "vertical": "s",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "toggleTermsModal": [Function],
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          "type": [Function],
        },
      ],
      "type": Symbol(react.fragment),
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
